shards/shards.go

```diff
@@ -150,17 +153,97 @@ type shardedSearcher struct {
 	ranked        []rankedShard
 
 	priority map[string]float64
+
+	// public is a set of public repository names.
+	public map[string]struct{}
 }
 
 func newShardedSearcher(n int64) *shardedSearcher {
 	ss := &shardedSearcher{
 		shards:   make(map[string]rankedShard),
 		sched:    newScheduler(n),
 		priority: make(map[string]float64),
+		public:   make(map[string]struct{}),
 	}
 	return ss
 }
 
+// readPublicSet reads public.txt and acquires an exclusive lock to hydrate
+// shardedSearcher with a set of public repos.
+func (ss *shardedSearcher) readPublicSet(dir string, lastMtime time.Time) time.Time {
+	publicPath := filepath.Join(dir, "public.txt")
+
+	st, err := os.Stat(publicPath)
+	if err != nil {
+		log.Printf("watchPublic: os.Stat: %s", err)
+		return lastMtime // ignore missing file errors
+	}
+	if st.ModTime() == lastMtime {
+		log.Printf("watchPublic: public.txt unchanged, no need to reload")
+		return lastMtime // file is unchanged
+	}
+
+	file, err := os.Open(publicPath)
+	if err != nil {
+		log.Printf("watchPublic: os.Open: %s", err)
+		return st.ModTime()
+	}
+	defer file.Close()
+
+	var publicRepos []string
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		repo := strings.TrimSpace(scanner.Text())
+		if repo == "" {
+			continue
+		}
+		publicRepos = append(publicRepos, repo)
+	}
+	if err := scanner.Err(); err != nil {
+		log.Printf("watchPublic: scanner error: %s", err)
+		return st.ModTime()
+	}
+
+	log.Printf("Reloading public.txt. %d public repositories", len(publicRepos))
+
+	publicReposSet := make(map[string]struct{}, len(publicRepos))
+	for _, r := range publicRepos {
+		publicReposSet[r] = struct{}{}
+	}
+	proc := ss.sched.Exclusive()
+	defer proc.Release()
+
+	// We invalidate the cache (ss.ranked = nil) to ensure that indexData.public is
+	// set for the next search. Instead we could also keep the cache intact and
+	// mutate ss.shards, but cache invalidation seems more explicit and doesn't rely
+	// on the mutated field having a reference like value.
+	ss.public = publicReposSet
+	ss.rankedVersion++
+	ss.ranked = nil
+
+	return st.ModTime()
+}
+
+// watchPublic watches public.txt in dir by calling readPublicSet in regular
+// intervals. The first call to readPublicSet is immediate.
+func (ss *shardedSearcher) watchPublic(dir string, done chan struct{}) {
```

Comment:
I know we already decided on this but I forgot the reason so here it goes again: Why don't we write the visibility bit to the indexData at index time? It's easy to make the change to get visibility data from Sourcegraph's API.

Reply:
We kind of do that. At index-time we update public.txt and index new repositories. However we keep these two things separate.

We don't want to store visibility as part of the index (on disk) because we want to keep the shard immutable. On the other hand it is very convenient to have the information in indexData when we create matchtrees.  Hence we have to hydrate indexData whenever we load the shard or the visibility changes. 

> It's easy to make the change to get visibility data from Sourcegraph's API.

We already get the visibility data from Sourcegraph's API. Zoekt asks frontend regularly what to index. The visibility data is part of the payload.

Reply:
> We don't want to store visibility as part of the index (on disk) because we want to keep the shard immutable.

I expect visibility data to be less frequently changing than the actual contents of a repo, which cause a re-index anyways?

Reply:
I agree, visibility should rarely change. It is more of a design decicion about where any metadata should live which changes independently of the indexed code (stars, forks, visibility, ...). So far we decided to keep those things in separate files on disk (loosely inspired by Lucene).

Reply:
I quite like the idea of storing it as part of the on-disk shard. We went back and forth on this idea a bit. The main downside for it being part of the shard is we "resurrect" old shards during rollouts. This means it's not that unlikely for a repo that went from public to private to briefly be available in the public index again during a rollout. However, we could work around that for sure.  Additionally it isn't the worst thing that code that was once public has a brief blip of being available shortly after it went private. It might make things _much_ simpler to change this design, and now is the perfect time to do it. Lets chat about it today again and finalize the decision!

FYI one benefit of this decision is it will make zoekt being user perm aware easier. We can change the decision of what is visible per request. However, I don't think this benefit should weigh into the decision since it's easy enough to introduce query.RepoFunc when we actually need it.

Reply:
I don't know. IMO it is not desireable to store rapidly changing metaData (for example stars)  in the shard. How do we treat updates? Tombstone the repo and reindex?